report/4.Solution/GeneratedCppCode.tex

```diff
@@ -0,0 +1,87 @@
+% !TeX root = ../main.tex
+
+\subsection{Generated C++ code}
+
+While the previous section primarily focused on the code generation from an abstract point of view, this section will go into depth about the generated C++ code.
+Since C++ is the output language of the HCL compiler, all the language concepts of HCL has to somehow be represented in valid C++ code.
+
+\textbf{Primitives}
+
+The primitives of HCL includes numbers, text and booleans. 
+Booleans are straightforward, as HCL's boolean behaves exactly like the C++ boolean type.
+This means that the regular boolean type can be used for output type of booleans.
+
+As HCL aims for simplicity, only one number type was introduced. 
+However this type should also be able to store non-integers. 
+Therefore the simple solution was to turn all HCL numbers into C++'s double type.
+The solution with doubles is not ideal, but it will fulfill the requirements of the number type in HCL.
+
+C++ does have an implementation of a string type in its standard library.
+However, the Arduino does not include this implementation.
+The Arduino uses its own string implementation that varies from the standard implementation.
+Along with this, the string type shares a lot of functionality with that of lists.
+Therefore the representation of strings in the generated C++ code, is based on the ConstList that is explained it the list section below.
+
+\textbf{Lists}
+
+As for lists, C++ includes a list type in its standard library call vector. 
+However, this vector type is yet again not available in the default libraries provided by the Arduino.
+Furthermore the vector implementation is mutable.
+Since the lists in HCL is immutable, a simple implementation of a constant list was implemented.
+
+One thing to be aware of, when using lists, is that the size cannot always be determined at compile time.
+This means that lists will usually do heap allocation. 
+With heap allocation comes the need to do proper memory handling, eg.
+clean up allocated memory in order to avoid memory leaks.
+
+This is done by utilizing the C++ concept of RAII\footnote{Resource acquisition is initialization} along with C++ smart pointers.
+All this really means, is that a pointer is implemented that keeps a count of the references to the memory pointed to.
+If the counter goes to zero, this means that there are no more references to memory, and the memory may be deallocated safely.
+
+It should be noted, that the current HCL compiler uses the std::shared\_ptr, which is not available in the Arduino standard library.
+The implementation should however be easily ported to the Arduino when needed.
```

Comment:
(Maybe) add a little bit of text explaining how the implementation can be ported.